sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -97,12 +99,38 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
       options = Map.empty)
   }
 
-  protected def getExternalTmpPath(
+  // Mostly copied from Context.java#getMRTmpPath of Hive 2.3.
+  // Visible for testing.
+  private[execution] def getNonBlobTmpPath(
+      hadoopConf: Configuration,
+      sessionScratchDir: String,
+      scratchDir: String): Path = {
+
+    // Hive's getMRTmpPath uses nonLocalScratchPath + '-mr-10000',
+    // which is ruled by 'hive.exec.scratchdir' including file system.
+    // This is the same as Spark's #oldVersionExternalTempPath.
+    // Only difference between #oldVersionExternalTempPath and Hive 2.3.0's is HIVE-7090.
+    // HIVE-7090 added user_name/session_id on top of 'hive.exec.scratchdir'
+    // Here it uses session_path unless it's emtpy, otherwise uses scratchDir.
+    val sessionPath = if (!sessionScratchDir.isEmpty) sessionScratchDir else scratchDir
+    val mrScratchDir = oldVersionExternalTempPath(new Path(sessionPath), hadoopConf, sessionPath)
```

Comment:
`oldVersionExternalTempPath` will use the first `Path` parameter's scheme to create the temp path.

Does it guarantee `Path(sessionPath)` always uses HDFS scheme here?

If `sessionPath` is a HDFS path, it is ok. But if we use `scratchDir` as `sessionPath` here, it will be something like `Path("/tmp/hive")`, is it still HDFS scheme?

Reply:
Hive has two kinds of scratch dir accordingly, one in local, the other in hdfs.
https://mingyue.me/2018/11/17/hive-scratch-working-directory/

In this pull-request, the latter one, `hive.exec.scratchdir` is used. In my recognition, we can assume HDFS schema here.

Reply:
But as you see we obtain scratch dir like `val scratchDir = hadoopConf.get("hive.exec.scratchdir", "/tmp/hive")`, it is possible that we use `"/tmp/hive"`. I think in the case we will use local scheme as the temp path.

Reply:
Here, no specific scheme is configured in `hive.exec.scratchdir` (just `/tmp/hive` without scheme). It means that the default scheme determined by `fs.default.name` is used. 
If `fs.default.name` is a local scheme like `file://` then `file://` will be used. If `fs.default.name` is a hdfs scheme like `hdfs://host:port/` then `hdfs://host:port/` will be used. I confirmed this behavior in my end to make sure it.

In most cases, `fs.default.name` is a hdfs scheme. 
Even if local scheme is used here, it should work because we can assume local scheme can be used in such environment (e.g. unit testing).

Reply:
When this new feature is enabled, it is possible that a scheme which doesn't work for this feature is used, e.g. local scheme. If it is happened and causes some error, end-users might not know how to deal with it.

Because we don't know if every scheme supports this feature, we use a list of schemes as config value, instead of a boolean config. Similarly, I think we should not reply on an assumption that `fs.default.name` always works for this feature. Can we just restrict this feature to HDFS only?

Reply:
Currently we just rely on `hive.exec.scratchdir`, and it works in most use cases even if `hive.exec.scratchdir` is not configured explicitly. 
I do not want to restrict this feature to HDFS only because I have seen some clusters which do not have HDFS. I want to let end-users choose any scheme where they want to store temporary data.

Reply:
If any one scheme doesn't work? Do we have clear error message to tell users what happens?

Reply:
It makes sense to me.Thank you for the suggestion.
I added an error message to suggest end-users to configure `hive.exec.scratchdir` if the scheme of the temporary directory is `file`.
Can you check if it is clear enough for end-users?

```
    val path = new Path(mrScratchDir, "-mr-10000")
    val scheme = Option(path.toUri.getScheme).getOrElse("")
    if (schema == "file") {
      logWarning(s"Temporary data will be written into a local disk " +
        s"(scheme: '$scheme', path: '$mrScratchDir'). " +
        s"You need to configure 'hive.exec.scratchdir' to use accessible location " +
        s"(e.g. HDFS path) from any executors in the cluster.")
    }
```